io.github.TTimo.GtkRadiant.json

```diff
@@ -0,0 +1,220 @@
+{
+  "app-id": "io.github.TTimo.GtkRadiant",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.28",
+  "sdk": "org.gnome.Sdk",
+  "command": "gtkradiant",
+  "finish-args": [
+    "--share=network",
+    "--share=ipc",
+    "--socket=x11",
+    "--socket=pulseaudio",
+    "--device=dri",
+    "--filesystem=host"
```

Comment:
why host filesystem permissions?

Reply:
This kind of application requires to read files from installation folder and read and write files from and to user folder of games that are not installed with Flatpak and that existed before Flatpak was even an idea. The purpose is to load the data of a game, to create new content for the game, to write the new content in the place the game expects it, then to load it from the game.

Reply:
and? `host` is VERY BROAD. And might not even do what you think it does.

Reply:
> and? `host` is VERY BROAD. And might not even do what you think it does.

I know enough of flatpak to even know `host` is not broad enough for this application.

Let's give an example. Let's pick the [Quake 3 Arena](https://en.wikipedia.org/wiki/Quake_III_Arena) game which was released in 1999, and for which the file layout and file path conventions for reading and writing were written in stone in 1999 and could not be modified without breaking compatibility.

With some luck Quake 3 would install in `/opt` and `host` would be broad enough. With bad luck it would install in `/usr` and then you would not be able to use this GtkRadiant flatpak to read content for this game because `host` is not broad enough.

Then, once you managed to read the game data with the GtkRadiant editor, you start creating content for the game (basically, game levels). The Quake 3 game expects the new content files produced by GtkRadiant to be written by GtkRadiant in `${HOME}/.q3a/baseq3` which is the user home folder of the game. There is no other way to do it, this is defined by the game and the game cannot read files from other places.

GtkRadiant is the official in-house id Software editor for Quake 3 and Quake Live, ported to GTK2 by them and made open source by them decades ago.

GtkRadiant is a level editor for many games developed in the last 20 years, starting with Quake 3 which was released in 1999 to games like [Unvanquished](https://unvanquished.net/) which is still in active development in 2022 (14 games are actually listed in GtkRadiant, including well known franchises like Wolfenstein, Star Trek, Jedi Knight and others, and this number becomes way larger when you start counting the countless mods based on them).

Just to give a glance on how much old the history of this software is, and how written in stone some design may be, GtkRadiant inherits some design from QuakeEd which was a NextStep application for ancestors of Quake 3 that was first released in 1996 and existed internally at id Software before that. . Changing some of those design requires to take a time machine and to jump 26 or 27 years ago in time, or if you can't afford a time machine, to rewrite all the binaries that were published and installed in millions of computers since more than 20 years, we talk about some of the most popular games in history.

So yes some of those design really don't feel like 2022 ones, but this is history and legacy.

So, this kind of editor should read and write files in places that are dictated by other applications, GtkRadiant is not the application deciding where to read and write. Reading and writing elsewhere would be a GtkRadiant bug making the application unusable.

Then, a big part of the design this application has to comply with is that the games it edits files for are implementing an user space file system and this application has to load the whole game file system when editing one file. So for example loading a file with a file picker that would open a portal to just the file being opened cannot work: you would need access to the whole game user file system to be able to edit the file, or  even create a brand new file without even opening it with a file picker first.

This application doesn't read and write files, it read and write an other application's user file system. This user file system isn't an image file, it's a whole tree of folders and files, and GtkRadiant requires arbitrary access to all of those folders and files to be able to click the “create” button (not even the “open” and “save” one).

Another thing that is worth mentioning is that not only GtkRadiant ships a graphical application, but GtkRadiant ships also some dedicated compilers, though those compilers don't compile C++ code, they compile 3D levels. Such compilers are also meant to be called from outside the flatpak when building game source directories. I know no one who would question the requirement for a compiler to have a broad access to a filesystem when called by a build system. Though, while one can work around that issue with some `flatpak run` option when calling the  compiler, the application needs to run the compiler with broad access.

On that compiler point of view, GtkRadiant is a bit like an IDE that would require it's compiler to access system's `/usr/include`, (that said, that just reminds me some [flatpak problems](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1504) users cannot even work around with `--filesystem=host` because of some flatpak design shortcoming).

Note: I'm not a maintainer of GtkRadiant itself though I'm a contributor to GtkRadiant. I'm myself a maintainer of a fork named [NetRadiant](https://netradiant.gitlab.io/) (exchanging code back and forth with GtkRadiant) and I also plan to make a flatpak for NetRadiant. So not only I know well the needs of GtkRadiant, but I also scrutinize with much attention and interest any answer you will do because like GtkRadiant, NetRadiant inherits the same legacy design no one has any power to change.

Outside of the topic of “old libraries and old SDK being used”, if GtkRadiant fails to be included into flathub, what would block it for inclusion would probably block NetRadiant as well. This would be the same for other applications like [DarkRadiant](https://www.darkradiant.net/), [UFO Radiant](https://ufoai.org/wiki/UFORadiant_installation), [NRC](https://github.com/Garux/netradiant-custom), [worldspawn](https://github.com/VeraVisions/worldspawn), and [others], plus some other editors not sharing GtkRadiant code like [J.A.C.K](https://jack.hlfx.ru/en/main.html) or [TrenchBroom](https://trenchbroom.github.io/). If the usage of the `--filesystem=host` option would block GtkRadiant for inclusion this would mean flathub/flatpak would not be compatible with more-than 25 years old design and method of working of the industry.

Wait… TrenchBoom is actually [included in flathub as a flatpak](https://flathub.org/apps/details/com.kristianduske.TrenchBroom) and… [it sets `--filesystem=host`](https://github.com/flathub/com.kristianduske.TrenchBroom/blob/4c17169f220f323ee5910f27ca010dac9af352d3/com.kristianduske.TrenchBroom.json#L14).